src/1.generic-constraints/index.ts

```diff
@@ -4,7 +4,17 @@
 // You can remove the readonly modifier from the mapped type by using -readonly
 
 // ------- start coding here -------
-
+function mapObjectProperties<
+  S extends Record<string, any>,
+  M extends Record<string, keyof S>
+>(source: S, mapping: M): { -readonly [K in keyof M]: S[M[K]] } {
```

Comment:
Return type nya masih salah. Harusnya yang tidak di-mapping jgn dihilangkan.

Untuk data type transformedUser2 harusnya:
```
const transformedUser2: {
    id: number;
    name: string;
    userAge: number;
}
```

Sedangkan disini hasilnya:
```
const transformedUser2: {
    id: number;
    name: string;
}
```

Menurut saya untuk merapihkan kode, lebih baik return type nya dipisah seperti ini:
```
type Mapping<S, M extends Record<string, keyof S>> = {
  -readonly [K in keyof M]: S[M[K]];
} & Omit<S, M[keyof M]>;
```

Disini kita gabungkan dengan `Omit<S, M[keyof M]>` supaya jika tidak ada di mapping maka propertinya tidak akan hilang.

Reply:
saya coba perbaiki, namun masih ada error di baris 18, dimana key nya is not assignable to type
<img width="723" alt="Screenshot 2024-09-23 at 11 55 22" src="https://github.com/user-attachments/assets/abaaa1d8-6380-4f2e-8055-dec302d7dafb">

lalu di baris 62, masih belum berhasil untuk produce error untuk transformedUser3

mohon saran dan arahannya, terima kasih

Reply:
Kalau kita gunakan kode berikut:
```
for (const key in mapping) {
    const sourceKey = mapping[key];
    result[key] = source[sourceKey];
  }
```

Maka kita hanya akan membuat object baru menggunakan value dari property yang ada di mapping object. Artinya original property nya akan hilang kalau tidak ada di list property.

Yang benar adalah:
```
type Mapping<S, M extends Record<string, keyof S>> = {
  -readonly [K in keyof M]: S[M[K]];
} & Omit<S, M[keyof M]>;

function mapObjectProperties<
  S extends Record<string, any>,
  M extends Record<string, keyof S>
>(source: S, mapping: M): Mapping<S, M> {
  const result = { ...source } as any;

  for (const key in mapping) {
    const sourceKey = mapping[key];
    result[key] = source[sourceKey];
    delete result[sourceKey];
  }

  return result;
}
```

Notice disini kita menggunakan as casting `{ ...source } as any;`
Alasannya karena keynya ada yang kita buang ada yang kita tambah. Dan kita sebagai programmer lebih tahu daripada typescript. Sebetulnya return typenya itu sudah selesai sejak awal, yang perlu kita awasi itu adalah kode javascript kita benar atau tidak.

Kode berikut sudah menyelesaikan masalah return type (kita menggunakan mapped type untuk membuat return type):
```
type Mapping<S, M extends Record<string, keyof S>> = {
  -readonly [K in keyof M]: S[M[K]];
} & Omit<S, M[keyof M]>;
```

Jadi kita sudah tidak usah pusingin lagi, langsung saja kita pakai type casting (as any):
```
const result = { ...source } as any;
```

Disini kita udah ga ketergantungan lagi dengan typescript untuk membuat tipe data berdasarkan kode javascript dalam function.

Tapi kita sebagai programmer juga harus pastikan apakah original data type nya sesuai dengan return type yang kita buat. Karena kalau kita pakai kode berikut:
```
for (const key in mapping) {
    const sourceKey = mapping[key];
    result[key] = source[sourceKey];
  }
```

Ini tidak sebanding dengan return type yang sudah kita buat, tapi akan tetap lolos karena kita pakai type casting `as any`, jadi disini kita harus berhati-hati. Karena untuk sesuatu yang complex seperti transformation, kita kadang ga ketergantungan lagi dengan tipe data yang dibuat oleh typescript berdasarkan javascript code, tapi langsung kita buat return typenya secara manual.

Reply:
Saya errornya keluar kok. Errornya bukan di variable transformedUser3, tapi di argument ke 2 functionnya.

![CleanShot 2024-09-23 at 18 55 20@2x](https://github.com/user-attachments/assets/663c2aba-6671-4aee-93ef-230092303a84)

Reply:
oh error yang itu, ok paham koh, terima kasih